examples/platform/esp32/mode-support/static-supported-modes-manager.cpp

```diff
@@ -0,0 +1,187 @@
+                        ModeOptionsProvider(nullptr, nullptr));
+    ESP32Config::Key countKey(ESP32Config::kConfigNamespace_ChipFactory, keyBuf);
+    VerifyOrReturnValue(ESP32Config::ReadConfigValue(countKey, supportedModeCount) == CHIP_NO_ERROR,
+                        ModeOptionsProvider(nullptr, nullptr));
+
+    modeOptionStructList = new ModeOptionStructType[supportedModeCount];
+    if (modeOptionStructList == nullptr)
+    {
+        return ModeOptionsProvider(nullptr, nullptr);
+    }
+
+    for (int index = 0; index < supportedModeCount; index++)
+    {
+        Structs::ModeOptionStruct::Type option;
+        uint32_t supportedModeMode = 0;
+        uint32_t semanticTagCount  = 0;
+        size_t outLen              = 0;
+
+        memset(keyBuf, 0, sizeof(char) * ESP32Config::kMaxConfigKeyNameLength);
+        VerifyOrReturnValue(ESP32Config::KeyAllocator::SupportedModesLabel(keyBuf, sizeof(keyBuf), endpointId, index) ==
+                                CHIP_NO_ERROR,
+                            ModeOptionsProvider(nullptr, nullptr), CleanUp());
+        ESP32Config::Key labelKey(ESP32Config::kConfigNamespace_ChipFactory, keyBuf);
+        VerifyOrReturnValue(ESP32Config::ReadConfigValueStr(labelKey, nullptr, 0, outLen) == CHIP_NO_ERROR,
+                            ModeOptionsProvider(nullptr, nullptr), CleanUp());
+
+        char * modeLabel = new char[outLen + 1];
+        VerifyOrReturnValue(ESP32Config::ReadConfigValueStr(labelKey, modeLabel, outLen + 1, outLen) == CHIP_NO_ERROR,
+                            ModeOptionsProvider(nullptr, nullptr), CleanUp());
+
+        memset(keyBuf, 0, sizeof(char) * ESP32Config::kMaxConfigKeyNameLength);
+        VerifyOrReturnValue(ESP32Config::KeyAllocator::SupportedModesValue(keyBuf, sizeof(keyBuf), endpointId, index) ==
+                                CHIP_NO_ERROR,
+                            ModeOptionsProvider(nullptr, nullptr), CleanUp());
+        ESP32Config::Key modeKey(ESP32Config::kConfigNamespace_ChipFactory, keyBuf);
+        VerifyOrReturnValue(ESP32Config::ReadConfigValue(labelKey, supportedModeMode) == CHIP_NO_ERROR,
+                            ModeOptionsProvider(nullptr, nullptr), CleanUp());
+
+        memset(keyBuf, 0, sizeof(char) * ESP32Config::kMaxConfigKeyNameLength);
+        VerifyOrReturnValue(ESP32Config::KeyAllocator::SemanticTagsCount(keyBuf, sizeof(keyBuf), endpointId, index) ==
+                                CHIP_NO_ERROR,
+                            ModeOptionsProvider(nullptr, nullptr), CleanUp());
+        ESP32Config::Key stCountKey(ESP32Config::kConfigNamespace_ChipFactory, keyBuf);
+        VerifyOrReturnValue(ESP32Config::ReadConfigValue(stCountKey, semanticTagCount) == CHIP_NO_ERROR,
+                            ModeOptionsProvider(nullptr, nullptr), CleanUp());
+
+        semanticTags = new SemanticTag[semanticTagCount];
+        if (semanticTags == nullptr)
+        {
+            CleanUp();
```

Comment:
This does not actually clean up sanely:

1. It deletes mode information for other endpoints, which is a strange thing to do just because we failed to allocate things for _this_ endpoint.
2. It leaks all the things we have allocated so far during this function call, since those are not actually attached to any entry in `epModeOptionsProviderList` yet and `CleanUp` only frees things in `epModeOptionsProviderList`.

Separately, why do we have a null-check here but not for the allocations of modeLabel?

Reply:
Updated to delete current endpoint only. Also added null check for `modeLabel`